### DISCUSSION TITLE
Tesla API Token Questions

### DISCUSSION THREAD

**Discussion by zipzagster** [alandtse/tesla#132] [Q&A]
1. Is there a way to use 2FA on my Tesla account and still be able to generate tokens?
2. What is the expiration time of the tokens? The token generating app I use has a reminder feature that doesn't match the actual expiration.
3. What is the practical difference between updating an expired token and using a new token instead of a refresh token?

Thanks

**Comment by alandtse** [comment]
1. Yes.
2. 45 days or whatever Tesla wants. If you're referring to the mass reset this week, Tesla invalidated everyone's token due to a security issue.
3. Depends on if Tesla invalidates old refresh tokens but there might be an upper limit on how many can be active. For more info on what this is see https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/

**Reply by zipzagster** [reply]
Thanks. What app will use 2FA and generate a token? I haven't turned 2FA on yet because I hear people complaining. Anyone successfully using refresh tokens?

**Reply by alandtse** [reply]
Tesla Tokens on Android works fine with 2FA for me.
